provider/core/src/filter/impls.rs

```diff
@@ -0,0 +1,227 @@
+    /// use icu_provider::filter::LanguageIdentifierFilter;
+    /// use icu_locid_macros::langid;
+    ///
+    /// let allowlist = vec![langid!("de"), langid!("zh")];
+    /// let provider = HelloWorldProvider::new_with_placeholder_data()
+    ///     .filter_by_langid_allowlist_strict(&allowlist);
+    ///
+    /// // German requests should succeed:
+    /// let req_de = DataRequest {
+    ///     resource_path: ResourcePath {
+    ///         key: key::HELLO_WORLD_V1,
+    ///         options: langid!("de").into(),
+    ///     }
+    /// };
+    /// let response: Result<DataResponse<HelloWorldV1Marker>, _> =
+    ///     provider.load_payload(&req_de);
+    /// assert!(matches!(response, Ok(_)));
+    ///
+    /// // English requests should fail:
+    /// let req_en = DataRequest {
+    ///     resource_path: ResourcePath {
+    ///         key: key::HELLO_WORLD_V1,
+    ///         options: langid!("en-US").into(),
+    ///     }
+    /// };
+    /// let response: Result<DataResponse<HelloWorldV1Marker>, _> =
+    ///     provider.load_payload(&req_en);
+    /// assert!(matches!(response, Err(DataError::FilteredResource(_, _))));
+    /// assert_eq!(
+    ///     "Resource was filtered: Locale filter (allowlist: [de, zh]): icu4x/helloworld@1/en-US",
+    ///     response.unwrap_err().to_string()
+    /// );
+    /// ```
+    fn filter_by_langid_allowlist_strict(
+        self,
+        allowlist: &'a [LanguageIdentifier],
+    ) -> RequestFilterDataProvider<Self, Self::F>;
+
+    /// Require that data requests contain a langid.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_provider::prelude::*;
+    /// use icu_provider::hello_world::*;
+    /// use icu_provider::filter::LanguageIdentifierFilter;
+    /// use icu_locid_macros::langid;
+    ///
+    /// let provider = HelloWorldProvider::new_with_placeholder_data()
+    ///     .require_langid();
+    ///
+    /// // Requests with a langid should succeed:
+    /// let req_with_langid = DataRequest {
+    ///     resource_path: ResourcePath {
+    ///         key: key::HELLO_WORLD_V1,
+    ///         options: langid!("de").into(),
+    ///     }
+    /// };
+    /// let response: Result<DataResponse<HelloWorldV1Marker>, _> =
+    ///     provider.load_payload(&req_with_langid);
+    /// assert!(matches!(response, Ok(_)));
+    ///
+    /// // Requests without a langid should fail:
+    /// let req_no_langid = DataRequest {
+    ///     resource_path: ResourcePath {
+    ///         key: key::HELLO_WORLD_V1,
+    ///         options: Default::default(),
+    ///     }
+    /// };
+    /// let response: Result<DataResponse<HelloWorldV1Marker>, _> =
+    ///     provider.load_payload(&req_no_langid);
+    /// assert!(matches!(response, Err(DataError::FilteredResource(_, _))));
+    /// ```
+    fn require_langid(self) -> RequestFilterDataProvider<Self, Self::F>;
+}
+
+impl<'a, T> LanguageIdentifierFilter<'a> for T {
+    type F = Box<dyn Fn(&DataRequest) -> bool + 'a>;
```

Comment:
I think the more idiomatic way of doing this would actually be to have 


```rust
trait LIFilter {
   fn filter(&DataRequest) -> bool;
}

struct IdFilter<F: Fn(&LanguageIdentifier)>(F);
impl LIFilter for IdFilter { ... }
...
```

And then have constructors that produce a `RequestFilterDataProvider<T, IdFilter>`, etc. This is similar to how iterators work.

But if that's too much, i think just moving these methods to be constructors on RFDP would be fine

Reply:
I did it this way because:

1. The primary purpose of the trait is to create syntax that allows you to chain filters: `icu_testdata::get_provider().require_locale().filter_locales(...).filter_keys(...)`
2. It keeps the `RequestFilterDataProvider` as a "pure", non-opinionated struct, and the canned implementations can be pulled in separately
3. The locale filtering functions can be grouped on one trait, key filtering on another, variant filtering on another, etc.

In particular, I wanted to avoid having a bunch of different structs that implement a filtering trait.  The proposed design has just one general RequestFilterDataProvider that forwards everything through a predicate function.  Then, the only job is to figure out the most ergonomic way to generate the predicate function.

Reply:
I'd be open to the idea of replacing the predicate function with a trait that has a single function like "filter", as you suggested with `LIFilter`.  Can you discuss the pros and cons of a predicate function (`Fn`) versus a single-function trait?

Reply:
Does the filtering stuff look better to you now?

Reply:
Yep!